packages/ra-core/src/controller/saveContext/useMutationMiddlewares.ts

```diff
@@ -0,0 +1,64 @@
+import { useCallback, useMemo, useRef } from 'react';
+
+/**
+ * Internal hook used to handle middlewares.
+ */
+export const useMutationMiddlewares = <
+    MutateFunc extends (...args: any[]) => any = (...args: any[]) => any
+>(): UseMutationMiddlewaresResult<MutateFunc> => {
+    const callbacks = useRef<Middleware<MutateFunc>[]>([]);
+
+    const addMiddleware = useCallback((callback: Middleware<MutateFunc>) => {
+        callbacks.current.push(callback);
+    }, []);
+
+    const removeMiddleware = useCallback((callback: Middleware<MutateFunc>) => {
+        callbacks.current = callbacks.current.filter(cb => cb !== callback);
+    }, []);
+
+    const getMutateWithMiddlewares = useCallback((fn: MutateFunc) => {
+        return (...args: Parameters<MutateFunc>) => {
+            let index = callbacks.current.length - 1;
+
+            const next = (...newArgs: any) => {
```

Comment:
![image](https://user-images.githubusercontent.com/99944/168824864-39f0ab6a-ecfb-4bd9-af6c-dfe1837e7865.png)

This magic code is a bot hard to reread, would you mind adding a bit of comments?